src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
+                        return;
+                    }
+
+                    _state = StreamState.Aborted;
 
-                    _responseComplete = true;
-                    _responseAborted = true;
+                    if (_waiterTaskSource != null)
+                    {
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
+                    }
+                }
 
-                    if (!_responseHeadersAvailable.Task.IsCompleted)
+                if (waiterTaskSource != null)
+                {
+                    waiterTaskSource.SetResult(true);
+                }
+            }
+
+            private (Task waiterTask, bool isEmptyResponse) TryEnsureHeaders()
+            {
+                lock (SyncObject)
+                {
+                    if (_disposed)
                     {
-                        // We are still waiting for response headers, so fail that task
-                        _responseHeadersAvailable.SetException(new IOException(SR.net_http_invalid_response));
+                        throw new ObjectDisposedException(nameof(Http2Stream));
+                    }
 
-                        // We shouldn't be waiting on data, since we haven't processed headers yet
-                        Debug.Assert(_responseDataAvailable == null);
+                    if (_state == StreamState.Aborted)
+                    {
+                        throw new IOException(SR.net_http_invalid_response);
+                    }
+                    else if (_state == StreamState.ExpectingHeaders)
+                    {
+                        Debug.Assert(_waiterTaskSource == null);
+                        _waiterTaskSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        return (_waiterTaskSource.Task, false);
+                    }
+                    else if (_state == StreamState.ExpectingData)
+                    {
+                        return (null, false);
                     }
                     else
                     {
-                        if (_responseDataAvailable != null)
-                        {
-                            _responseDataAvailable.SetException(new IOException(SR.net_http_invalid_response));
-                            _responseDataAvailable = null;
-                        }
+                        Debug.Assert(_state == StreamState.Complete);
+                        return (null, _responseBuffer.ActiveSpan.Length == 0);
                     }
                 }
             }
 
+            public async Task ReadResponseHeadersAsync()
+            {
+                // Wait for response headers to be read.
+                Task waiterTask;
+                bool emptyResponse;
+                (waiterTask, emptyResponse) = TryEnsureHeaders();
+                if (waiterTask != null)
+                {
+                    await waiterTask;
```

Comment:
.ConfigureAwait(false)

Reply:
I intentionally leave these off just to test you.

Reply:
Makes sense, keep me on my toes, like pop quizzes in grade school.